src/deque.rs

```diff
@@ -811,6 +811,189 @@ impl<T, S: VecStorage<T> + ?Sized> DequeInner<T, S> {
     }
+
+    /// Shortens the deque, keeping the first `len` elements and dropping
+    /// the rest.
+    ///
+    /// If `len` is greater or equal to the deque's current length, this has
+    /// no effect.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use heapless::Deque;
+    ///
+    /// let mut buf: Deque<_, 5> = Deque::new();
+    /// buf.push_back(5);
+    /// buf.push_back(10);
+    /// buf.push_back(15);
+    /// buf.truncate(1);
+    /// assert_eq!(buf.make_contiguous(), [5]);
+    /// ```
+    pub fn truncate(&mut self, len: usize) {
+        /// Runs the destructor for all items in the slice when it gets dropped (gracefully or
+        /// during unwinding).
+        struct Dropper<'a, T>(&'a mut [T]);
+
+        impl<'a, T> Drop for Dropper<'a, T> {
+            fn drop(&mut self) {
+                unsafe {
+                    ptr::drop_in_place(self.0);
+                }
+            }
+        }
+
+        unsafe {
+            // If new desired length is greater or equal, we don't need to act.
+            if len >= self.storage_len() {
+                return;
+            }
+
+            let (front_idx, back_idx) = (self.front, self.back);
+            let (front, back) = self.as_mut_slices();
+
+            // If `len` desires to keep elements past front's entire length,
+            // then only back's contents need to be dropped
+            // as the two slices combined should be more than `len`.
+            if len > front.len() {
+                let begin = len - front.len();
+                let drop_back = back.get_unchecked_mut(begin..) as *mut _;
+
+                // When we have a non-contiguous deque,
+                // we find the new wrapped index for `back` using the desired length
+                self.back = self.to_physical_index(len);
+                self.full = false;
+
+                ptr::drop_in_place(drop_back);
+            } else {
+                // Otherwise, we know back's entire contents need to be dropped,
+                // since the desired length never reaches into it.
+                let drop_back = back as *mut _;
+                let drop_front = front.get_unchecked_mut(len..) as *mut _;
+
+                if back_idx <= front_idx {
+                    // Self::to_physical_index returns the index `len` units _after_ the front cursor,
+                    // meaning we also can use it to determine where the cap for front needs to be placed.
+                    self.back = self.to_physical_index(len);
+                } else {
+                    // If the back cursor is "in front" of the front cursor,
+                    // we can just move it closer to truncate front's length.
+                    let back_trunc = front.len() - len;
+
+                    // This shouldn't underflow since `back` must be larger
+                    // than both `len` and `front` for this to even be reached.
+                    self.back -= back_trunc;
+                }
```

Comment:
I don't understand why we can't use `to_physical_index` in both cases

Reply:
Ah, good catch. No reason why we can't, I just didn't reconsider that logic. Fixed!